### DISCUSSION TITLE
Failed to resolve module specifier "@vue/reactivity"

### DISCUSSION THREAD

**Discussion by simonferndriger** [alpinejs/alpine#3578] [1. Help]
I apologize in advance because this is probably not related to alpinejs. Since today, when I run the live preview (using the VS Code extension from Microsoft), I get this error in the console:

`Uncaught TypeError: Failed to resolve module specifier "@vue/reactivity". Relative references must start with either "/", "./", or "../".`

Does anyone know how to fix this? I didn't change anything since it worked last.

**Comment by ekwoka** [comment]
How are you using Alpinejs?

This last update included Alpine changing to do what all tools do and not pre-bundle dependencies.

**Comment by simonferndriger** [comment]
Since I'm currently still in developing state, I'm using the latest versions:

**import Alpine           from '//cdn.jsdelivr.net/npm/alpinejs@x.x.x/dist/module.esm.min.js';
import AlpineCollapse   from '//cdn.jsdelivr.net/npm/@alpinejs/collapse@x.x.x/dist/module.esm.min.js';
import AlpineI18n       from '//cdn.jsdelivr.net/npm/alpinejs-i18n@x.x.x/dist/module.esm.min.js';
import AlpineIntersect  from '//cdn.jsdelivr.net/npm/@alpinejs/intersect@x.x.x/dist/module.esm.min.js';
import AlpinePersist    from '//cdn.jsdelivr.net/npm/@alpinejs/persist@x.x.x/dist/module.esm.min.js';**

**Reply by ekwoka** [reply]
Okay easy solve

Just use https://esm.sh

**Reply by simonferndriger** [reply]
I'm now using Parcel to build the production version. However, it looks like Parcel can't handle the esm.sh imports. Any idea on how to work around this?

I already posted this issue on Parcel: https://github.com/parcel-bundler/parcel/discussions/9170

**Reply by ekwoka** [reply]
just do what you're supposed to do and install them locally to bundle. You can mark them as external which may work, or mark them as external with their normal names and add an import map for that.

Not sure why you aren't just installing them to your project if you're already installing other things...

**Comment by simonferndriger** [comment]
Awesome, it works now again with:

import Alpine           from '//esm.sh/alpinejs';
import AlpineCollapse   from ***@***.***/collapse';
import AlpineI18n       from '//esm.sh/alpinejs-i18n';
import AlpineIntersect  from ***@***.***/intersect';
import AlpinePersist    from ***@***.***/persist';

I'm just curious, why did it work before but not anymore the other way?

Am Di., 16. Mai 2023 um 18:14 Uhr schrieb Eric Kwoka <
***@***.***>:

**Reply by ekwoka** [reply]
In the past, AlpineJS was prebundling for the module. This is generally a bad practice (that's why almost no one does it).

With 3.12.1, the deps and no longer prebundled.

This normally isn't an issue but, on this case, Alpine depends on @vue/reactivity which is absolutely not able to be run in the browser on its normal module import.

So some of these module based cdns that don't do any processing, they would try to use the version that's not possible.

Esm.sh actually does process all the modules (with esbuild). You can even do a lot of modifications to the bundler process by adding query params to the url (like overwriting dependency versions, changing transpilation targets, and marking things as external).

So it's more like an actual bundling process.

**Comment by simonferndriger** [comment]
Wow. Thanks!

Do you perhaps have a tip for me such that I can switch easily to a fix version in production while using the latest version in dev?

**Reply by ekwoka** [reply]
Well, there isn't a `fixed` version at all.

Generally, if you are doing anything in production, you should NOT be using these public CDNs at all, and instead take the 2 seconds to bundle it for deployment.

**Comment by simonferndriger** [comment]
Sorry for doing it differently, but I'm not used to work with the console.
So I just use VS Code Live Preview and Parcel only to create an optimized
build...

Am Di., 1. Aug. 2023 um 10:27 Uhr schrieb Eric Kwoka <
***@***.***>:

**Comment by simonferndriger** [comment]
I tried to install them locally and run them also with Parcel, but then
when I try to import alpinejs as a local module, I get this error in the
console:
>
> Failed to load module script: Expected a JavaScript module script but the
> server responded with a MIME type of "text/html". Strict MIME type checking
> is enforced for module scripts per HTML spec.
>

Am Di., 1. Aug. 2023 um 12:21 Uhr schrieb Simon Ferndriger <
***@***.***>:

**Reply by ekwoka** [reply]
....in parcel or in the browser?

**Comment by simonferndriger** [comment]
In the browser. Parcel has no problems running...

[image: image.png]

Am Di., 1. Aug. 2023 um 13:16 Uhr schrieb Eric Kwoka <
***@***.***>:

**Reply by ekwoka** [reply]
Then it either isn't getting bundled, or the URI that Parcel is creating is invalid.

if you showed more of your code and looked at the details of the request giving this error, I'm sure you could figure it out.

Basically your javascript request is not returning javascript

**Comment by simonferndriger** [comment]
Hmm. Well, this is the entire website with the current esm.sh imports:
https://dev.swisswebcams.ch/_new/index.html

And this is the package.json:
{
  "dependencies": {
    "alpinejs": "^3.12.3",
    "parcel": "^2.9.3"
  },
  "devDependencies": {
    "parcel-bundler": "^1.12.5"
  },
  "name": "swc",
  "scripts": {
    "build": "parcel build --dist-dir rc/_new",
    "start": "parcel"
  },
  "source": "dev/_new/index.html",
  "version": "0.0.0"
}

Does that help?

Am Di., 1. Aug. 2023 um 13:33 Uhr schrieb Eric Kwoka <
***@***.***>:

**Reply by ekwoka** [reply]
Well, it works fine. So....

**Comment by simonferndriger** [comment]
Yes, it works with the esm.sh imports. But when I use the local
installation imports instead and run it with Parcel on http://localhost:1234,
the mentioned error occurs...

Am Di., 1. Aug. 2023 um 13:46 Uhr schrieb Eric Kwoka <
***@***.***>:

**Reply by ekwoka** [reply]
as in you did `npm i alpinejs` and `import Alpine from 'alpinejs'`?

What is the code parcel is outputting look like? Whats the request look like that is failing?

**Comment by simonferndriger** [comment]
Ah, yes I did. When I run the install command again I get:
>
> up to date, audited 1131 packages in 11s
>
> 158 packages are looking for funding
>   run `npm fund` for details
>
> 30 vulnerabilities (16 moderate, 14 high)
>
> To address issues that do not require attention, run:
>   npm audit fix
>
> To address all issues (including breaking changes), run:
>   npm audit fix --force
>
> Run `npm audit` for details.
>

And when I use import Alpine from 'alpinejs'; and run Parcel with npm run
start the result is:

> > ***@***.*** start
> > parcel

Server running at http://localhost:1234
> ✨ Built in 733ms


Unfortunately, I don't see any failed request in the network tab and don't
know where to look for otherwise, there is nothing in the Parcel console
neither:

[image: image.png]


Am Di., 1. Aug. 2023 um 14:33 Uhr schrieb Eric Kwoka <
***@***.***>:

**Comment by simonferndriger** [comment]
I now use Vite and it works all just perfectly :) One minor detail I need
to figure out. It does not yet see/include lazy loaded css files like this
(for the build process, in dev mode these files are just there and working
of course):

      <template x-if="view === 'map'">
        <link href="styles/map.css" rel="stylesheet">
      </template>

Any idea how to also include these in the minifying/build process by vite
(without losing the lazy load effect)?


Best
Simon

Am Di., 1. Aug. 2023 um 16:26 Uhr schrieb Simon Ferndriger <
***@***.***>:

**Reply by ekwoka** [reply]
You can do css lazy loading by doing something like

```
<link href="path" media="print" onload="this.media='all'; this.onload=null" rel="stylesheet" />
```

Basically, by default, browsers start downloading alternative stylesheets after the document is fully loaded, and don't block on them. Then it will trigger the onload and make it apply to everything.

You can also just put css link tags before the content that depends on it, and it'll only block rendering at that point in the document, which can give a nice cascading loading effect. This SHOULD be detected by Vite, but any further discussion on that would need to be at Vite's repo for handling those kinds of "out of place" link tags.

**Comment by simonferndriger** [comment]
Thank you Eric, that works for me as well :)

Am Do., 3. Aug. 2023 um 06:15 Uhr schrieb Eric Kwoka <
***@***.***>:
